reference/src/representation/enums.md

```diff
@@ -0,0 +1,376 @@
+
+**Example.** The following enum uses an `repr(u8)` annotation:
+
+```rust
+#[repr(u8)]
+enum TwoCases {
+    A(u8, u16),
+    B(u16),
+}
+```
+
+This will be laid out equivalently to the following more 
+complex Rust types:
+
+```
+union TwoCasesRepr {
+    A: TwoCasesVariantA,
+    B: TwoCasesVariantB,
+}
+        
+#[repr(u8)]
+enum TwoCasesTag { A, B }
+
+#[repr(C)]
+struct TwoCasesVariantA(TwoCasesTag, u8, u16);
+
+#[repr(C)]
+struct TwoCasesVariantB(TwoCasesTag, u16);
+```
+
+Note that the `TwoCasesVariantA` and `TwoCasesVariantB` structs are
+`#[repr(C)]`; this is needed to ensure that the `TwoCasesTag` value
+appears at offset 0 in both cases, so that we can read it to determine
+the current variant.
+
+#### C-compatible layouts.
+
+When the `#[repr]` tag includes `C`, e.g., `#[repr(C)]` or `#[repr(C,
+u8)]`, the layout of enums is changed to better match C++ enums. In
+this mode, the data is laid out as a tuple of `(discriminant, union)`,
+where `union` represents a C union of all the possible variants. The
+type of the discriminant will be the integral type specified (`u8`,
+etc) -- if no type is specified, then the compiler will select one
+based on what a size a C-like enum would have with the same number of
+variants.
+
+This layout, while more compatible and arguably more obvious, is also
+less efficient than the non-C compatible layout in some cases in terms
+of total size.
+
+**Example.** The following enum:
+
+```rust
+#[repr(C, Int)]
+enum MyEnum {
+    A(u32),
+    B(f32, u64),
+    C { x: u32, y: u8 },
+    D,
+}
+```
+
+is equivalent to the following Rust definition:
+
+```rust
+#[repr(C)]
+struct MyEnumRepr {
+    tag: MyEnumTag,
+    payload: MyEnumPayload,
+}
+
+#[repr(Int)]
+enum MyEnumTag { A, B, C, D }
+
+#[repr(C)]
+union MyEnumPayload {
+   A: u32,
+   B: MyEnumPayloadB,
+   C: MyEnumPayloadC,
+   D: (),
+}
+
+#[repr(C)]
+struct MyEnumPayloadB(f32, u64);
+
+#[repr(C)]
+struct MyEnumPayloadC { x: u32, y: u8 }
+}
+```
+
+This enum can also be represented in C++ as follows:
+
+```c++
+#include <stdint.h>
+
+enum class MyEnumTag: CppEquivalentOfInt { A, B, C, D };
+struct MyEnumPayloadB { float _0; uint64_t _1;  };
+struct MyEnumPayloadC { uint32_t x; uint8_t y; };
+
+union MyEnumPayload {
+   uint32_t A;
+   MyEnumPayloadB B;
+   MyEnumPayloadC C;
+};
+
+struct MyEnum {
+    MyEnumTag tag;
+    MyEnumPayload payload;
+};
+```
+
+## Enums without a specified representation
+
+If no explicit `#[repr]` attribute is used, then the layout of most
+enums is not specified, with one crucial exception: option-like enums
+may in some cases use a compact layout that is identical to their
+payload.
+
+(Meta-note: The content in this section is not described by any RFC
+and is therefore "non-normative".)
+
+### Discriminant elision on Option-like enums
+
+**Definition.** An **option-like enum** is an enum which has:
+
+- one variant with a single field,
+- other variants with no fields ("unit" variants).
```

Comment:
This definition is surprising to me. I don't recall any demand for that particular generalization and off-hand I can't think of a case where it would be really useful to have it guaranteed, and it requires a bunch of subtlety below that wouldn't otherwise be there.

Reply:
Yeah, why does this allow multiple other variants with no fields?

Reply:
It doesn't seem to require any subtlety in particular? But I'm fine to limit to one unit variant.

Reply:
Perhaps not terribly subtle in the grand scheme of things, but the vague talk about "an adequate number of niche values" and the third example could be cut, and more drastically it would become feasible to not introduce the "niche" terminology at all at this stage (just enumerate the types we special-case in the compiler today).